app/services/notification_generator.rb

```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+# Creates notifications for each approval that is expiring within 30 days. Clears
+# expiring notifications and any notifications that have received updated approvals
+class NotificationGenerator
+  def call
+    approvals_without_notification = Approval.where.missing(:notifications).where('expires_on between ? and ?',
+                                                                                  0.days.after,
+                                                                                  30.days.after)
+    if approvals_without_notification.length.positive?
+      approvals_without_notification.each do |record|
+        record.children.each { |child| generate_notification(record.id, child.id) }
+      end
+    end
+
+    delete_notifications
+  end
+
+  private
+
+  def delete_notifications
+    Approval.all.each do |record|
```

Comment:
Rather than looping through all Approvals, does it make sense to reduce the number of Approvals you're checking with a join?  `Approval.joins(:notifications).each`?

Reply:
I wasn't sure if the updated approvals shared the same ID. The approvals that I utilized in the test didn't share id and this join was problematic in the fact that the preexisting notification always linked back to the non updated approval, and I needed to check all the most recent approvals.